compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenDialect.cpp

```diff
@@ -52,41 +52,177 @@ void IREECodegenDialect::initialize() {
   addTypes<IREE::Codegen::NullPointerType>();
 }
 
+static LogicalResult validateTuningEntrypoint(
+    ModuleOp moduleOp, transform::NamedSequenceOp &kernelConfigOp,
+    int &numTuningEntryPoints, const std::string &kRequiredByDefaultAttr) {
+  for (Operation &op : moduleOp.getBody()->getOperations()) {
+    if (auto namedSeqOp = dyn_cast<transform::NamedSequenceOp>(&op)) {
+      if (namedSeqOp.getName() == kKernelConfigSpecName) {
+        kernelConfigOp = namedSeqOp;
+      }
+    }
+
+    if (op.hasAttr(kTuningSpecEntrypointAttrName)) {
+      ++numTuningEntryPoints;
+    }
+  }
+
+  if (!kernelConfigOp) {
+    return moduleOp->emitError()
+           << "Missing named sequence '" << kKernelConfigSpecName << "'"
+           << kRequiredByDefaultAttr;
+  }
+
+  // Verify that the kernelConfigOp has the attribute
+  // `iree_codegen.tuning_spec_entrypoint`.
+  if (!kernelConfigOp->hasAttr(kTuningSpecEntrypointAttrName)) {
+    return kernelConfigOp.emitError()
+           << "Missing attribute '" << kTuningSpecEntrypointAttrName
+           << "' in named sequence '" << kKernelConfigSpecName << "'"
+           << kRequiredByDefaultAttr;
+  }
+
+  if (numTuningEntryPoints != 1) {
+    return moduleOp.emitError()
+           << "Expected one named sequence with '"
+           << kTuningSpecEntrypointAttrName << "', but found "
+           << numTuningEntryPoints << kRequiredByDefaultAttr;
+  }
+
+  return success();
+}
+
+LogicalResult
+validateKernelConfigContents(transform::NamedSequenceOp kernelConfigOp,
+                             const std::string &kRequiredByDefaultAttr) {
+  transform::ForeachMatchOp foreachMatchOp;
+  bool hasYieldOp = false;
+  for (Block &block : kernelConfigOp.getBlocks()) {
```

Comment:
You can reduce the nesting here by erroring out when there's more than one block